docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,499 @@
+execution at all.
+
+### Concurrent Batches
+
+An alternative approach looks to gain concurrency support at a layer below SQL, in KV logic.
+If the restriction that only one `client.Batch` can execute in a `client.Txn` at a time was
+lifted then SQL statement execution logic would not have to change between serially executed
+statements and concurrently executed statements. Each independent statement could run in its own
+goroutine and use it's own `client.Batch`
+
+This offers a number of benefits including that it is easier to reason about from the SQL
+layer, it eliminates the issue about statements that execute multiple batches, it is
+a generally useful improvement to the KV client API that could be beneficial elsewhere,
+and it removes the restriction of [coalesced batches](#coalesced-batches) that statements
+need to access the KV layer in lockstep. This last point means that this alternative should
+also perform better.
+
+However, it is unclear how difficult lifting this restriction would be. At the very least,
+it would require `client.Txn` to manage its state in a thread-safe manner, and would also
+require changes to `kv.TxnCoordSender`.
+
+## Error Handling
+
+Error handling is a major concern with the batching of SQL statements. Regardless of whether
+statements are batched together or not, their error semantics should behave as if they were
+executed serially. This has a few consequences that can be demonstrated using an example of
+two ordered statements S<sub>1</sub> and S<sub>2</sub> that are being run in parallel. If during
+parallel execution we see an error for S<sub>2</sub>, we must still wait for all earlier
+statements (in this case, just S<sub>1</sub>) to finish. Only then can we determine which error
+to return, which should always correspond to the "earliest" error from the ordered statements
+that are being run in parallel. Likewise, if during execution we see an error for S<sub>1</sub>,
+we should attempt to actively cancel the execution of all later statements. We must also
+be sure that no statements go into effect if any of the parallel-executing statements throw
+an error. Conveniently, the atomicity property of transactions means that we can simply abort
+the transaction and achieve this behavior.
+
+If we make the feature opt-in, then it may be ok to relax these constraints and simplify
+the error handling. For instance, we may be able to loosen the guarantee that the earliest
+error seen in a parallel batch will be the one returned. This is not an unreasonable relaxation
+because parallel batches will only contain independent statements, and therefore the successful
+execution of one statement should never depend on the result of another earlier in the transaction
+but within the same batch. Still, it could be confusing to users to fix an error in one statement,
+only to find that there was also an error in an earlier statement.
```

Comment:
To be fair, given that the statements are indeed independent, they'll just have the opposite experience -- they'll fix the error in t
he earlier statement only to find that there was also an error in a later statement.